ls1intum/Artemis
# `General`: Only show next date in exercise header

### Checklist
#### General


#### Client

### Motivation and Context


The exercise header could get quite large on the right side, when all dates of the exercise lifecycle were used. This PR simplifies the header and moves the rest of the information to a table below the problem statement.

### Description

Only the due date and up to one more date is now shown in the exercise header. The other date is the next lying in the future, or if that is not the case the latest in the past that is after the due date. (ExampleSolutionPublicationDate is never chosen as it is not relevant enough).
All exercise dates are now displayed in a table below the problem statement. When there are no dates, some standard information is always shown (i.e. `Due Date: No Due Date`) instead of leaving the table blank.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- Different Exercises (programming, modeling, quiz)

1. Log in to Artemis
2. Create different exercises with different configurations or change the dates of existing ones and check that the next relevant date is shown in the header and all dates are shown in the table.
3. Especially check an exercise with no dates, only dates in the past, with a start date

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exam with an exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Check that the header displays in the assessment dashboard also shows the correct date

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage



| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [complaint.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4922-9/artifact/shared/Coverage-Report-Client-Tests/app/complaints/complaint.service.ts.html) | 89.74% | ✅ |
| [header-exercise-page-with-details.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4922-9/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-headers/header-exercise-page-with-details.component.ts.html) | 98.71% | ✅ |
| [exercise-info.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4922-9/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-info/exercise-info.component.ts.html) | 100% | ✅ |

### Screenshots


Header:
![Bildschirmfoto 2023-05-22 um 23 35 11](https://github.com/ls1intum/Artemis/assets/38322605/bd26e6e7-c29c-4698-9e1a-7dfa9bf026ac)
![Bildschirmfoto 2023-06-09 um 11 29 55](https://github.com/ls1intum/Artemis/assets/38322605/618d8a6a-44a7-439e-8884-8ff5af70df34)

Date table with varying window widths:
![Bildschirmfoto 2023-05-22 um 23 35 23](https://github.com/ls1intum/Artemis/assets/38322605/dc2ce1ce-efe7-4880-a5bf-e6945427ec87)
![Bildschirmfoto 2023-05-22 um 23 35 36](https://github.com/ls1intum/Artemis/assets/38322605/86ea832d-3012-4923-b73e-4e4f90101ada)
